### ISSUE TITLE
Unsupported operand types using += with floats initialized to 0 (not 0.0)

### ISSUE THREAD

**Issue by timabbott** [JukkaL/mypy#1193]
mypy complains with:
`/home/tabbott/foo.py:4: error: Unsupported operand types for + ("int" and "float")` with the following code:
```
foo = 0
foo += 0.5
```

While one can work around this by initializing to 0.0, it seems worth thinking about if mypy should accept this.

**Comment by gvanrossum** [comment]
I agree that we should do something about this. The question is, what
should the scope of this type adjustment be? What if I have a foo that
initializes a class variable foo to 0 and then a method executes
`self.foo += 0.5` ? What if that method is in a subclass?

The fix we have in place for variables initialized to empty lists only
works for locals -- but it is often insufficient, so I think we need a
more general architecture for these types of type adjustments.

**Comment by JukkaL** [comment]
We could probably tweak partial types to work for the original example, but as Guido pointed out, it would not work in general. However, it might be good enough. I'd like to see more real-world examples before deciding what to do with this.

Note that you can also add an annotation (though changing the initializer is usually easier):

```
foo = 0  # type: float
foo += 0.5
```

**Comment by timabbott** [comment]
Yep makes sense; for a data point I had to do this 6 times in zulip.git.


        -Tim Abbott

**Comment by JukkaL** [comment]
@timabbott Could you give more information about the 6 cases in zulip.git -- or were they all similar to `x = 0` / `x += 0.5`?

**Comment by timabbott** [comment]
See https://github.com/timabbott/zulip/commit/bf82ac3b2261f0335789e35868f287accb6cf525

5/6 are like that and the other is like that except x is in a dict.

**Comment by JelleZijlstra** [comment]
On this sample mypy currently emits `main.py:2: error: Incompatible types in assignment (expression has type "float", variable has type "int")  [assignment]`, which is a much more reasonable error than that in the original.

**Comment by hauntsaninja** [comment]
Yeah, and mypy allows it with `--allow-redefinition-new`
